lightningd/plugin.c

```diff
@@ -269,32 +296,54 @@ static bool plugin_read_json_one(struct plugin *plugin)
 		return false;
 	}
 
+	jrtok = json_get_member(plugin->buffer, toks, "jsonrpc");
+	resulttok = json_get_member(plugin->buffer, toks, "result");
+	errortok = json_get_member(plugin->buffer, toks, "error");
 	idtok = json_get_member(plugin->buffer, toks, "id");
 
-	if (!idtok) {
-		plugin_kill(plugin, "JSON-RPC response does not contain an \"id\"-field");
+	if (!jrtok) {
+		plugin_kill(
+		    plugin,
+		    "JSON-RPC message does not contain \"jsonrpc\" field");
 		return false;
 	}
 
-	/* We only send u64 ids, so if this fails it's a critical error (note
-	 * that this also works if id is inside a JSON string!). */
-	if (!json_to_u64(plugin->buffer, idtok, &id)) {
-		plugin_kill(plugin, "JSON-RPC response \"id\"-field is not a u64");
-		return false;
-	}
-
-	request = uintmap_get(&plugin->plugins->pending_requests, id);
-
-	if (!request) {
-		plugin_kill(plugin, "Received a JSON-RPC response for non-existent request");
-		return false;
+	if (idtok && (errortok || resulttok)) {
+		/* When a rpc call is made, the Server MUST reply with
```

Comment:
Remove all these checks, IMHO.  It's none of our business what they send back and forth, as long as it's an object with matching `id`.  That's minimal.

In fact, this patch seems entirely extraneous; the next patch should just touch the "!idtok" case....

Reply:
That might be true for passthrough RPC calls, but not all are passthrough, so we better make sure that they adhere to the JSON-RPC standard. We have `getmanifest` and `init` which are handled by us internally, and we can check either here, or in the callbacks, but this looked like a good place to do so.

Also we'll add more of the internally handled RPC calls, once we start moving the bitcoin connection to a plugin (`bitcoind`, neutrino, ...) so I'd like to keep them here.

Reply:
It's not just a layering violation, it's also slow for large objects; at least with 'id' it's likely to be at the front.  Especially once you add looking for "params" and "method" in the next patch.

We really should be doing this only once we've identified that it's something we need to deal with.  That's both minimal and clearer.